src/transport/SecureSessionMgr.cpp

```diff
@@ -207,6 +207,11 @@ CHIP_ERROR SecureSessionMgr::SendMessage(SecureSessionHandle session, PayloadHea
     if (bufferRetainSlot != nullptr)
     {
         encryptedMsg = msgBuf.Retain();
+        // Rewind the start and len of the buffer back to pre-send state for following possible retransmition.
+        encryptedMsg->SetStart(msgStart);
+        encryptedMsg->SetDataLength(msgLen);
```

Comment:
We should not rewind the buffer back to pre-send state before we complete the transmission.